### PR TITLE
Add authorization to the fetch spaces endpoint

### PR DESCRIPTION
**Pull request by danail-branekov**
## Is there a related GitHub Issue?
https://github.com/cloudfoundry/cf-k8s-controllers/issues/150

## What is this change about?
Add authorization to the `GET /v3/spaces` endpoint. User credentials are expected to be put into the `Authorization` request header.

## Does this PR introduce a breaking change?
Sort of, the create space (`POST /v3/spaces`) now implicitly requires authentication

## Acceptance Steps
See [the story](https://github.com/cloudfoundry/cf-k8s-controllers/issues/150) for acceptance

## Tag your pair, your PM, and/or team
@mnitchev

### CODE REVIEW

File: api/apis/space_handler.go [hunk 1]
```diff
@@ -87,8 +101,24 @@ func (h *SpaceHandler) SpaceListHandler(w http.ResponseWriter, r *http.Request)
 	orgUIDs := parseCommaSeparatedList(r.URL.Query().Get("organization_guids"))
 	names := parseCommaSeparatedList(r.URL.Query().Get("names"))
 
-	spaces, err := h.spaceRepo.FetchSpaces(ctx, orgUIDs, names)
+	spaceRepo, err := h.spaceRepoProvider.SpaceRepoForRequest(r)
+	if err != nil {
+		if authorization.IsUnauthorized(err) {
```
Comment (tcdowney) [inline api/apis/space_handler.go:106]: Question for my own understanding:

This just checks that the user submitted an auth header right? If they have correctly submitted an `Authorization` header it will get past this and if they lack the requisite RBAC then the call on line 119 to `FetchSpaces` will fail?

If that's the case, I'm wondering if instead of an unknown error we should check to see if the request failed due to authorization reasons and return a `403` instead in that case.

**Review by tcdowney — DISMISSED** [review]

File: api/apis/apis_suite_test.go [hunk 2]
```diff
@@ -114,6 +114,18 @@ func expectUnknownKeyError() {
     }`)
 }
 
+func expectUnauthorizedError() {
+	expectJSONResponse(http.StatusUnauthorized, `{
+        "errors": [
+            {
+                "detail": "No auth token was given, but authentication is required for this endpoint",
```
[the remark this answers is not in the document]
Comment (kieron-dev) [reply api/apis/apis_suite_test.go:121]: We're changing this to "Authentication error" as suggested.

File: api/apis/space_handler.go (hunk 1 above)
Comment (danail-branekov) [reply api/apis/space_handler.go:106]: > ... and if they lack the requisite RBAC then the call on line 119 to FetchSpaces will fail?
Not really. In order to list authorised spaces efficiently we lean on the privileged client to get them all in one shot and filter out the ones the authenticated user has been bound to. We cannot rely on RBAC as namespaces are cluster-wide resources, and therefore users (and the user-scoped client) are not expected to have permissions to list them. 

TL;DR, the call on line 119 cannot fail due to authorisation issues because that line would use the privileged client that can do anything.

Comment (tcdowney) [reply api/apis/space_handler.go:106]: Ah yeah, I see.

I guess that's more of an issue for the future endpoints then. Thanks for the clarification!

**Comment by mnitchev** [comment]
We've resolved the propagation issue. We decided to make the propagation of roles configurable in the role mappings. This means that roles like the OrgAuditor, OrgUser and OrgBillingManager that don't have permissions to list spaces within an org will not be propagated down to the space namespace, hence fixing the listing problem.

If there is no more feedback that needs to be addressed, this PR can be merged

**Review by georgethebeatle — APPROVED** [review]

**Comment by georgethebeatle** [comment]
@tcdowney can you please approve this PR so that we can merge. I can see that the conversation is resolved, so I am guessing you just forgot to click the button.
